indexd/index/drivers/query/urls.py

```diff
@@ -0,0 +1,75 @@
+    "postgresql": dict(array_agg=func.array_agg, string_agg=func.string_agg)
+}
+
+
+class AlchemyURLsQueryDriver(URLsQueryDriver):
+    """SQLAlchemy based impl"""
+
+    def __init__(self, alchemy_driver):
+        """ Queries index records based on URL
+        Args:
+            alchemy_driver (indexd.index.drivers.alchemy.SQLAlchemyIndexDriver):
+        """
+        self.driver = alchemy_driver
+
+    def query_urls(self, exclude=None, include=None, only_versioned=True, offset=0, limit=1000):
+
+        with self.driver.session as sxn:
+            # special database specific functions dependent of the selected dialect
+            q_func = driver_query_map.get(sxn.bind.dialect.name)
+
+            query = sxn.query(IndexRecordUrl.did, q_func['string_agg'](IndexRecordUrl.url, ",").label("urls"))\
+                .outerjoin(IndexRecord)
+
+            # filter by version
+            if only_versioned is True:
+                query = query.filter(IndexRecord.version.isnot(None))
+            elif only_versioned is False:
+                query = query.filter(~IndexRecord.version.isnot(None))
+
+            query = query.group_by(IndexRecordUrl.did)
+
+            # add url filters
+            if include and exclude:
+                query = query.having(and_(~q_func['string_agg'](IndexRecordUrl.url, ",").contains(exclude),
+                                     q_func['string_agg'](IndexRecordUrl.url, ",").contains(include)))
+            elif include:
+                query = query.having(q_func['string_agg'](IndexRecordUrl.url, ",").contains(include))
+            elif exclude:
+                query = query.having(~q_func['string_agg'](IndexRecordUrl.url, ",").contains(exclude))
+
+            # [('did', 'urls')]
+            record_list = query.order_by(IndexRecordUrl.did.asc()).offset(offset).limit(limit).all()
+        return record_list
+
+    def query_metadata_by_key(self, key, value, url=None, only_versioned=True, offset=0, limit=1000):
+
+        with self.driver.session as sxn:
+            query = sxn.query(IndexRecordUrlMetadata.did,
+                              IndexRecordUrlMetadata.url,
+                              IndexRecord.rev)\
+                .filter(IndexRecord.did == IndexRecordUrlMetadata.did) \
+                .filter(IndexRecordUrlMetadata.key == key) \
+                .filter(IndexRecordUrlMetadata.value == value)
+
+            # filter by version
+            if only_versioned is True:
+                query = query.filter(IndexRecord.version.isnot(None))
+            elif only_versioned is False:
+                query = query.filter(~IndexRecord.version.isnot(None))
+
+            # add url filter
+            if url:
+                query = query.filter(IndexRecordUrlMetadata.url.like("%{}%".format(url)))
+
+            # [('did', 'url', 'rev')]
+            record_list = query.order_by(IndexRecordUrlMetadata.did.asc()).offset(offset).limit(limit).all()
+        return record_list
```

Comment:
it would be helpful to change this to a dictionary with the identifiers as keys, or put it into its own object, or something else rather than using `record_list[0]` etc. which is hard to understand

Reply:
That would be nice, but sqlalchemy returns a list of sets, and so far I have not found away to get sqlalchemy to return a dict or custom object instead. In java I would dynamically transform the results to non entity objects. I also don't want to loop through the list here since another loop will be required to split the urls into an array and removed fields that should not be returned. I also find things like `result_list[0]` very confusing

Reply:
right, so here couldn't you just return a dictionary?
```python
return dict(zip(['did', 'url', 'rev'], record_list))
```

Reply:
`urls` is a list, that the query return as a comma delimited string from the query, so doing dict -> zip still means I have to split it or convert it to list. The best solution here is to get sqlalchemy to work on it. If I move the logic in here, it will essentially be the same logic as is currently

Reply:
yes, the thing is there's no reason to have it in this format because nothing is using it this way, only as a dictionary; it only gets used in `/metadata/q` which is doing the weird dictionary conversion right now, so it may as well be converted to a dictionary here, because it's clearer (one more note on this: you can also just `return [record.__dict__ for record in record_list]` which is probably the shortest/simplest that gets)

which `urls` are you talking about? I thought `IndexRecordUrlMetadata` had just the single `url` field?

Reply:
I can't view it

Reply:
Expected response for for endpoints is like this:
```
[
 {"did": "AAAA-BB", "urls": ["s3://some-randomly-awesome-url"], "rev": "XXXX"},
 {"did": "AAAA-CC", "urls": ["s3://another-randomly-awesome-url"], "rev": "XXYX"}
]
```
Fields in this response can be removed based on the `fields` request parameter. 
SqlAlchemy returns a list of tuples. In `query_urls` the urls with that tuple is a comma separated list, in `query_urls_metadata` it started out as comma separated list of all the urls, but was reduced to just the url matching the pattern. To removed undesired columns and make the urls a list, I do one loop (which I am not happy with, but can't find an alternative using only sqlalchemy).
You are suggesting I do an initial loop, to convert result_list to a dict automatically. I can do that, but it would not cater for removing undesired fields and converting urls to a list. Meaning I either loop through the dict again, to do this or skip the dict conversion you suggested and bring in the current logic for removing the undesired fields and converting urls to list. If I do option 2, it will be the exact same solution as I currently have now and will still use something like result_lis[1]